AudioKit/Common/Nodes/Playback/Player/AKPlayer.swift

```diff
@@ -0,0 +1,724 @@
+
+    // MARK: - Loading
+
+    /// Replace the contents of the player with this url
+    public func load(url: URL) throws {
+        self.audioFile = try AVAudioFile(forReading: url)
+        initialize()
+    }
+
+    /// Mostly applicable to buffered players, this loads the buffer and gets it ready to play.
+    /// Otherwise it just sets the startTime and endTime
+    public func preroll(from startingTime: Double = 0, to endingTime: Double = 0) {
+        var from = startingTime
+        let to = endingTime
+
+        if from > to {
+            from = 0
+        }
+
+        startTime = from
+        endTime = to
+
+        Swift.print("preroll from \(startTime) to \(endTime)")
+
+        guard isBuffered else { return }
+        updateBuffer()
+    }
+
+    // MARK: - Playback
+
+    /// Play entire file right now
+    public func play() {
+        play(from: startTime, to: endTime, at: nil, hostTime: nil)
+    }
+
+    /// Play segments of a file
+    public func play(from startingTime: Double, to endingTime: Double = 0) {
+        var to = endingTime
+        if to == 0 {
+            to = endTime
+        }
+        play(from: startingTime, to: to, at: nil, hostTime: nil)
+    }
+
+    /// Play file using previously set startTime and endTime at some point in the future
+    public func play(at audioTime: AVAudioTime?) {
+        play(at: audioTime, hostTime: nil)
+    }
+
+    /// Play file using previously set startTime and endTime at some point in the future with a hostTime reference
+    public func play(at audioTime: AVAudioTime?, hostTime: UInt64?) {
+        play(from: startTime, to: endTime, at: audioTime, hostTime: hostTime)
+    }
+
+    /// Play file using previously set startTime and endTime at some point in the future specified in seconds
+    /// with a hostTime reference
+    public func play(when scheduledTime: Double, hostTime: UInt64?) {
+        play(from: startTime, to: endTime, when: scheduledTime, hostTime: hostTime)
+    }
+
+    public func play(from startingTime: Double,
+                     to endingTime: Double,
+                     when scheduledTime: Double,
+                     hostTime: UInt64? = nil) {
+        let refTime = hostTime != nil ? hostTime! : mach_absolute_time()
+        let avTime = AKPlayer.secondsToAVAudioTime(hostTime: refTime, time: scheduledTime)
+        play(from: startingTime, to: endingTime, at: avTime, hostTime: refTime)
+    }
+
+    /// Play using full options. Last in the convenience play chain, all play() commands will end up here
+    public func play(from startingTime: Double, to endingTime: Double, at audioTime: AVAudioTime?, hostTime: UInt64?) {
+        preroll(from: startingTime, to: endingTime)
+        schedule(at: audioTime)
+        playerNode.play()
+
+        completionTimer?.invalidate()
+        prerollTimer?.invalidate()
+
+        let prerollTime = audioTime != nil ?
+            AKPlayer.audioTimeToSeconds(hostTime: hostTime!, audioTime: audioTime!) : 0
```

Comment:
Here's a crasher.
There's nothing here to prevent optional hostTime from being force unwrapped while nil.  
Make hostTime non-optional, or guard where it'll be used.
Forced unwrapping should be avoided as a rule.

Reply:
good catch!